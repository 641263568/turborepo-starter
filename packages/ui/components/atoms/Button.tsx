import { forwardRef, MouseEvent } from "react";
import { classNames } from "../../helpers/classNames";

export interface ButtonProps {
  variant: "black" | "ghost" | "white" | "ghostWhite";
  label: string;
  onClick: (e: MouseEvent<HTMLButtonElement>) => unknown;
  disabled: boolean;
  icon?: any | SVGElement;
  iconAlign?: "left" | "right";
  type: "button" | "submit";
}

export type ButtonRef = HTMLButtonElement;

export const Button = forwardRef<ButtonRef, ButtonProps>((props, ref) => {
  const {
    variant = "black",
    label,
    onClick,
    disabled,
    icon,
    iconAlign,
    type = "button",
  } = props;

  return (
    <button
      ref={ref}
      type={type}
      onClick={(e) => onClick(e)}
      disabled={disabled}
      className={classNames(
        variant === "black"
          ? `text-white bg-black rounded ${
              disabled ? "text-black bg-grey4 rounded" : ""
            }`
          : "",
        variant === "white"
          ? `bg-white text-black rounded ${
              disabled ? "text-black bg-grey4 rounded" : ""
            }`
          : "",
        variant === "ghost"
          ? `bg-opacity-0 border border-black rounded ${
              disabled ? "text-black bg-grey4 rounded" : ""
            }`
          : "",
        variant === "ghostWhite"
          ? `text-white bg-black ${
              disabled ? "text-black bg-grey4 rounded" : ""
            }`
          : "",
        icon && iconAlign === "right" ? "flex-row-reverse" : "",
        "h-[3.125rem] w-[10.375rem] relative inline-flex items-center justify-center hover:bg-opacity-90 transition-all"
      )}
    >
      {icon && (
        <span
          className={classNames(
            icon && iconAlign === "left" && "mr-2",
            icon && iconAlign === "right" && "ml-2"
          )}
        >
          {icon}
        </span>
      )}
      {label}
      {variant === "ghostWhite" && (
        <span className="absolute border border-white border-solid inset-[1px] rounded" />
      )}
    </button>
  );
});

Button.displayName = "Button";
