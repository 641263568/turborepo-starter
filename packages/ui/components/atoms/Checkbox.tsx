import uniqueId from "lodash/uniqueId";
import { useState, useEffect, ChangeEvent, forwardRef } from "react";
import { classNames } from "../../helpers/classNames";

export interface CheckboxProps {
  label?: string | undefined;
  onChange: (e: ChangeEvent<HTMLInputElement>) => unknown;
  disabled?: boolean;
  checked?: boolean;
}

export type CheckboxRef = HTMLInputElement;

export const Checkbox = forwardRef<CheckboxRef, CheckboxProps>((props, ref) => {
  const { label, onChange, disabled, checked } = props;
  const [isChecked, setIsChecked] = useState<boolean | undefined>(checked);
  const [checkboxId, setCheckboxId] = useState<string>("");

  useEffect(() => {
    const id: string = uniqueId(
      label
        ? `${label.trim().replace(/\s+/g, "-").toLocaleLowerCase()}-`
        : uniqueId("checkbox-")
    );
    setCheckboxId(id);
  }, [label]);

  return (
    <span className={classNames(label ? "inline-flex items-center" : "")}>
      <input
        ref={ref}
        onClick={() => setIsChecked(!isChecked)}
        onChange={(e) => onChange(e)}
        disabled={disabled}
        checked={isChecked}
        type="checkbox"
        id={checkboxId}
        className={classNames(
          disabled
            ? "bg-grey3 border-grey3 cursor-not-allowed"
            : "border-black dark:border-grey2",
          isChecked ? "checked:border-white" : "",
          "cursor-pointer w-[0.938rem] h-[0.938rem] border-solid border rounded-none bg-white/0 focus:ring-black dark:focus:border-grey2 focus:ring-1 text-black"
        )}
        tabIndex={0}
      />
      {label && (
        <label
          className={classNames(
            label && disabled ? "text-grey3" : "",
            isChecked ? "dark:text-white" : "dark:text-grey2",
            "text-caption text-black ml-3"
          )}
          htmlFor={checkboxId}
        >
          {label}
        </label>
      )}
    </span>
  );
});

Checkbox.displayName = "Checkbox";
