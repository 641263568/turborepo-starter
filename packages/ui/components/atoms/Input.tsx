import * as React from "react";
import { forwardRef, ChangeEvent, useState, MouseEvent } from "react";
import { classNames } from "../../helpers/classNames";

export interface InputProps {
  placeholder: string;
  type?: "text" | "email" | "password" | string;
  disabled?: boolean;
  icon?: any | SVGElement;
  iconLabel?: string;
  required?: boolean;
  error?: boolean;
  errorMessage?: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => unknown;
  onButtonClick?: (e: MouseEvent<HTMLDivElement>) => unknown;
}

export type InputRef = HTMLInputElement;

export const Input = forwardRef<InputRef, InputProps>((props, ref) => {
  const {
    placeholder,
    type,
    disabled,
    icon,
    iconLabel,
    required,
    error,
    errorMessage,
    onChange,
    onButtonClick,
  } = props;
  const [value, setValue] = useState<string>("");

  const handleChangle = (e: ChangeEvent<HTMLInputElement>) =>
    setValue(e.target.value);

  const input = (
    <input
      ref={ref}
      type={type}
      placeholder={required ? "* " + placeholder : placeholder}
      aria-label={placeholder}
      onChange={(e) => {
        onChange(e), handleChangle(e);
      }}
      value={value}
      disabled={disabled}
      required={required}
      className={classNames(
        disabled
          ? "cursor-not-allowed border-b-grey3 placeholder:text-grey3"
          : "",
        error
          ? "text-error border-b-error placeholder:text-error"
          : "text-black border-b-grey3 focus:border-b-black placeholder:text-grey2",
        "text-body2 w-full border-x-0 border-t-0 border-b focus:ring-white/0 focus:outline-none h-[2.063rem] p-2.5 bg-white/0 transition-colors"
      )}
    />
  );

  if (icon && iconLabel) {
    return (
      <div className="relative">
        {input}
        <div
          className={classNames(
            error ? "text-error" : "text-black",
            "absolute inset-y-0 right-0 flex items-center cursor-pointer hover:opacity-90"
          )}
          role="button"
          tabIndex={0}
          onClick={onButtonClick}
        >
          <span
            className={classNames(
              error ? "text-error text-caption" : "text-black text-body2",
              "mr-[0.5rem]"
            )}
          >
            {error && errorMessage ? "* " + errorMessage : iconLabel}
          </span>
          {icon}
        </div>
      </div>
    );
  }

  if (icon) {
    return (
      <div className="relative">
        {input}
        <div
          className={classNames(
            error ? "text-error" : "text-black",
            "absolute inset-y-0 right-0 flex items-center cursor-pointer"
          )}
          role="button"
          tabIndex={0}
          onClick={onButtonClick}
        >
          <span
            className={classNames(
              error ? "text-error text-caption" : "text-black text-body2",
              "mr-[0.5rem]"
            )}
          >
            {error && errorMessage && "* " + errorMessage}
          </span>{" "}
          {icon}
        </div>
      </div>
    );
  }

  if (error && errorMessage) {
    return (
      <div className="relative">
        {input}
        <div className="absolute inset-y-0 right-0 flex items-center cursor-pointer text-error text-caption">
          {"* " + errorMessage}
        </div>
      </div>
    );
  }

  return input;
});

Input.displayName = "Input";
