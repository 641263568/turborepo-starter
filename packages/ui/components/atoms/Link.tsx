import { forwardRef } from "react";
import { classNames } from "../../helpers/classNames";

export interface LinkProps {}

export type LinkRef = HTMLDivElement;

export const Link = forwardRef<LinkRef, LinkProps>((props, ref) => {
  return <></>;
});

Link.displayName = "Link";
