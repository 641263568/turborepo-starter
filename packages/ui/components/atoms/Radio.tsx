import uniqueId from "lodash/uniqueId";
import { forwardRef, ChangeEvent } from "react";
import { classNames } from "../../helpers/classNames";

export interface RadioProps {
  type: string;
  legend?: string;
  label?: string;
  description?: string;
  options: {
    value?: string;
    checked?: boolean;
    disabled?: boolean;
  }[];
  variant?: "default" | "inline";
  onChange: (e: ChangeEvent<HTMLInputElement>) => unknown;
}

export type RadioRef = HTMLInputElement;

export const Radio = forwardRef<RadioRef, RadioProps>((props, ref) => {
  const { type, label, legend, description, options, variant, onChange } =
    props;

  return (
    <div>
      {label && (
        <label className="text-body1 text-black dark:text-white">{label}</label>
      )}
      {description && (
        <p className="text-body2 leading-5 text-grey2 mb-4">{description}</p>
      )}
      <fieldset>
        {legend && <legend className="sr-only">{legend}</legend>}
        <div
          className={classNames(
            variant === "default"
              ? "space-y-4"
              : "space-y-4 sm:flex sm:items-center sm:space-y-0 sm:space-x-10"
          )}
        >
          {options.map((option, index: number) => {
            const id: string = uniqueId("radio-");

            return (
              <div key={index} className="flex items-center">
                <input
                  onChange={(e) => onChange(e)}
                  ref={ref}
                  id={id}
                  name="radio"
                  type={type}
                  defaultChecked={option.checked}
                  className={classNames(
                    option.disabled
                      ? "border-grey3 cursor-not-allowed"
                      : "border-black dark:border-grey2 cursor-pointer",
                    "w-[0.938rem] h-[0.938rem"
                  )}
                  disabled={option.disabled}
                />
                {option.value && (
                  <label
                    htmlFor={id}
                    className="ml-3 block text-caption cursor-pointer text-black dark:text-white"
                  >
                    {option.value}
                  </label>
                )}
              </div>
            );
          })}
        </div>
      </fieldset>
    </div>
  );
});

Radio.displayName = "Radio";
