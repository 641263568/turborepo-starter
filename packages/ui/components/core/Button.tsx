import { forwardRef, MouseEvent } from "react";
import { classNames } from "../../helpers/classNames";

export enum ButtonVariant {
  Black = "Black",
  Ghost = "Ghost",
  White = "White",
  GhostWhite = "GhostWhite",
}

export interface ButtonProps {
  variant: ButtonVariant;
  label: string;
  onClick: (e: MouseEvent<HTMLButtonElement>) => unknown;
  disabled?: boolean;
  icon?: any | SVGElement;
  iconAlign?: "left" | "right";
  type?: "submit" | "reset" | "button" | undefined;
}

export type ButtonRef = HTMLButtonElement;

export const Button = forwardRef<ButtonRef, ButtonProps>((props, ref) => {
  const { label, onClick, disabled = false, type = "button" } = props;

  return (
    <button
      ref={ref}
      type={type}
      onClick={(e) => onClick(e)}
      disabled={disabled}
      className={classNames("")}
    >
      {label}
    </button>
  );
});

Button.displayName = "Button";
