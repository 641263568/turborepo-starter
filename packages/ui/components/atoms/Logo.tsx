import Link from "next/link";

export interface LogoImage {
  as: "image";
  alt: string;
  title: string;
  el: any;
}

export interface LogoText {
  as: "text";
  title: string;
}

export type LogoProps = LogoImage | LogoText;

export const Logo = (logo: LogoProps) => {
  if (logo.as === "text") {
    return (
      <div className="dark:text-white text-black">
        <Link href="/">{logo.title}</Link>
      </div>
    );
  }

  return (
    <Link href="/">
      <a className="dark:text-white text-black">
        {logo.el}
        <span className="sr-only">{logo.title}</span>
      </a>
    </Link>
  );
};

Logo.displayName = "Logo";
