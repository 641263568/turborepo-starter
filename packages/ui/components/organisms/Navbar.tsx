import Link from "next/link";
import { forwardRef, ReactNode } from "react";
import { classNames } from "../../helpers/classNames";
import { Logo } from "../atoms/Logo";

export interface NavbarItemEl {
  as: "el";
  el: ReactNode;
}

export interface NavbarItemLinks {
  as: "link";
  name: string;
  href: string;
  current: boolean;
}

export enum NavbarAlignItems {
  Left = "left",
  Right = "right",
}

export interface NavbarProps {
  alignItems?: NavbarAlignItems;
  items: Array<NavbarItemEl | NavbarItemLinks> | undefined;
}

export type NavbarRef = HTMLDivElement;

export const Navbar = forwardRef<NavbarRef, NavbarProps>((props, ref) => {
  const { items, alignItems } = props;
  const align = alignItems ? alignItems : NavbarAlignItems.Right;

  return (
    <header className="bg-background dark:bg-black" ref={ref}>
      <div className="mx-auto px-2 sm:px-6 lg:px-8 max-w-navbar">
        <div className="relative flex items-center h-navbar">
          <div className="grid grid-cols-12 flex-1 gap-grid">
            <div className="col-span-2 flex items-center">
              <Logo as="text" title="Emry" />
            </div>
            {items && (
              <div
                className={classNames(
                  align === NavbarAlignItems.Left
                    ? "col-start-4"
                    : "col-span-10 ml-auto"
                )}
              >
                {items && (
                  <nav>
                    <ul className="flex space-x-16 items-center">
                      {items.map((item, index) => {
                        if (item.as === "link") {
                          return (
                            <li
                              key={index}
                              className={classNames(
                                item.current
                                  ? "border-b border-b-grey3 dark:border-b-grey1"
                                  : "hover:-mb-[7px]",
                                "text-body2 pb-1.5 -mb-1.5 transition duration-300 hover:border-b hover:border-black dark:hover:border-white dark:text-white text-black"
                              )}
                            >
                              <Link
                                aria-current={item.current}
                                href={item.href}
                              >
                                {item.name}
                              </Link>
                            </li>
                          );
                        }

                        return <li key={index}>{item.el}</li>;
                      })}
                    </ul>
                  </nav>
                )}
              </div>
            )}
          </div>
        </div>
      </div>
    </header>
  );
});

Navbar.displayName = "Navbar";
