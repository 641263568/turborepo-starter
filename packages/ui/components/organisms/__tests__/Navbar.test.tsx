import { render } from "@testing-library/react";
import { Navbar, NavbarProps } from "../Navbar";
import { Button } from "../../atoms/Button";
import { axe, toHaveNoViolations } from "jest-axe";

expect.extend(toHaveNoViolations);

jest.mock("next/link", () => {
  return ({ children }: any) => {
    return children;
  };
});

const items: NavbarProps["items"] = [
  {
    as: "link",
    name: "Features",
    href: "/features",
    current: true,
  },
  {
    as: "link",
    name: "About",
    href: "/about",
    current: false,
  },
  {
    as: "link",
    name: "Contact",
    href: "/contact",
    current: false,
  },
  {
    as: "el",
    el: (
      <Button
        onClick={() => console.log("Button click")}
        variant="ghost"
        label="Download"
      />
    ),
  },
];

test("navbar passes a11y test", async () => {
  const { container } = render(<Navbar items={items} />);
  expect(await axe(container)).toHaveNoViolations();
});

it("navbar renders unchanged", () => {
  const { container } = render(<Navbar items={items} />);
  expect(container).toMatchSnapshot();
});
