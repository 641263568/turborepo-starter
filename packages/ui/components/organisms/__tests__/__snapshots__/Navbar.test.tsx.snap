// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navbar renders unchanged 1`] = `
<div>
  <header
    class="bg-background dark:bg-black"
  >
    <div
      class="mx-auto px-2 sm:px-6 lg:px-8 max-w-navbar"
    >
      <div
        class="relative flex items-center h-navbar"
      >
        <div
          class="grid grid-cols-12 flex-1 gap-grid"
        >
          <div
            class="col-span-2 flex items-center"
          >
            <div
              class="dark:text-white text-black"
            >
              <a
                href="/"
              >
                Emry
              </a>
            </div>
          </div>
          <div
            class="col-span-10 ml-auto"
          >
            <nav>
              <ul
                class="flex space-x-16 items-center"
              >
                <li
                  class="border-b border-b-grey3 dark:border-b-grey1 text-body2 pb-1.5 -mb-1.5 transition duration-300 hover:border-b hover:border-black dark:hover:border-white dark:text-white text-black"
                >
                  <a
                    href="/features"
                  >
                    Features
                  </a>
                </li>
                <li
                  class="hover:-mb-[7px] text-body2 pb-1.5 -mb-1.5 transition duration-300 hover:border-b hover:border-black dark:hover:border-white dark:text-white text-black"
                >
                  <a
                    href="/about"
                  >
                    About
                  </a>
                </li>
                <li
                  class="hover:-mb-[7px] text-body2 pb-1.5 -mb-1.5 transition duration-300 hover:border-b hover:border-black dark:hover:border-white dark:text-white text-black"
                >
                  <a
                    href="/contact"
                  >
                    Contact
                  </a>
                </li>
                <li>
                  <button
                    class="bg-opacity-0 border border-black dark:border-white dark:text-white text-black rounded  hover:bg-opacity-90 transition-all h-[3.125rem] w-[10.375rem] relative inline-flex items-center justify-center"
                    type="button"
                  >
                    Download
                  </button>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
